core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -47,43 +47,48 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
     with WorkspaceSupport
     with LazyLogging {
 
-  def createSnapshot(workspaceName: WorkspaceName, snapshot: NamedDataRepoSnapshot): Future[DataRepoSnapshotResource] =
+  def createSnapshot(workspaceName: WorkspaceName,
+                     snapshotIdentifiers: NamedDataRepoSnapshot
+  ): Future[DataRepoSnapshotResource] =
     getV2WorkspaceContextAndPermissions(workspaceName,
                                         SamWorkspaceActions.write,
                                         Some(WorkspaceAttributeSpecs(all = false))
-    ).flatMap { workspaceContext =>
-      val wsid = workspaceContext.workspaceIdAsUUID // to avoid UUID parsing multiple times
-      // create the stub workspace in WSM if it does not already exist
-      if (!workspaceStubExists(wsid, ctx)) {
-        workspaceManagerDAO.createWorkspace(wsid, ctx)
+    ).flatMap { rawlsWorkspace =>
+      val wsid = rawlsWorkspace.workspaceIdAsUUID // to avoid UUID parsing multiple times
+      val snapshot = dataRepoDAO.getSnapshot(snapshotIdentifiers.snapshotId, ctx.userInfo.accessToken)
+      val validator = new SnapshotReferenceCreationValidator(rawlsWorkspace, snapshot)
+      val aggregatedWorkspaceService = new AggregatedWorkspaceService(workspaceManagerDAO)
+
+      // prevent disallowed access across workspace or dataset protection boundaries
+      validator.validateProtectedStatus()
+
+      Try(aggregatedWorkspaceService.getAggregatedWorkspace(rawlsWorkspace, ctx)) match {
+        case Success(wsmWorkspace) =>
+          // if there's an existing WSM workspace, make sure its platform is compatible
+          // with that of the snapshot's dataset.
+          validator.validatePlatforms(wsmWorkspace.getCloudPlatform)
+        case Failure(_) =>
+          // Pre-emptively fail (and prevent creation of a workspace stub here if snapshot is Azure). We can remove this
+          // check once we support Azure-based snapshots by reference.
+          validator.validateSnapshotPlatform()
+
+          // create the stub workspace in WSM if it does not already exist
+          workspaceManagerDAO.createWorkspace(wsid, ctx)
```

Comment:
I think this call is misplaced. We need to call `workspaceManagerDAO.createWorkspace(wsid, ctx)` when:
1. this is a GCP workspace
2. there does not already exist a workspace record in WSM

if condition 1 above is true, then `aggregatedWorkspaceService.getAggregatedWorkspace(rawlsWorkspace, ctx)` will always return `Success` and we'll never reach this code.

Reply:
Just to clarify, but is your list an *and* or an *or* :)

I want to make sure I'm clear on whether or not this should be happening as a side effect for Azure workspaces (my hunch was no, and that the current behavior, which creates a workspace no matter what, isn't desirable).

Reply:
My list is an "and" - hopefully the comment I added just before this one helps to clarify. And you're correct, if this is an Azure workspace we should not make any attempt to create a stub in WSM.